### PR TITLE
Update K8s executor secret env var section

### LINKED ISSUE #2770: Workers on K8s Executor already have Astro env variables available
### Issue Type

Missing content for an existing feature

### Links to Affected Docs

https://docs.astronomer.io/astro/kubernetes-executor#mount-secret-environment-variables-to-worker-pods

### The Issue or Context

Astro environment variables are already available to Worker Pods in K8s Executor. [Slack thread](https://astronomer.slack.com/archives/C01MCEU9EV7/p1692895708008429).

### Fix or Proposed Change

_No response_

### Additional Notes

_No response_

### Required Reviewers

_No response_

### PR DESCRIPTION
**Pull request by manmeetkaur**
Resolves #2770 

Add a simple method to access env variables in addition to the complex method that requires `pod_override` in only special cases. Also change how env vars are stored on astro.

### CODE REVIEW

**Review by jwitz — CHANGES_REQUESTED** [review]
@manmeetkaur The KE and KPO have two different use cases, so I don't see this as redundant content. What if we just rewrite this section to have the standard `os.environ` usage? We don't have that documented anywhere else right now- to be honest we should have it documented for Celery too, or in your [env var docs](https://github.com/astronomer/docs/pull/2801), but a simple topic for KE would work for the time being.

**Comment by manmeetkaur** [comment]
> @manmeetkaur The KE and KPO have two different use cases, so I don't see this as redundant content. What if we just rewrite this section to have the standard `os.environ` usage? We don't have that documented anywhere else right now- to be honest we should have it documented for Celery too, or in your [env var docs](https://github.com/astronomer/docs/pull/2801), but a simple topic for KE would work for the time being.

@jwitz These are two same use-cases: using secrets in KPO; irrespective of if it is in KE or CE, it works the exactly the same. And this example is incorrect. Why do you want to keep the exact same thing in two places  when we have one generic doc for KPO  which is irrespective of the executor used ? 

`os.environ()` is the use in PythonOperator and is the standard across the python world, nothing special. **if** it needs to be added it should be in LEARN.

**Comment by jwitz** [comment]
@manmeetkaur These are not the exact same thing, because one uses the KPO and one doesn't. We're about to remove instructions that, even if they were inaccurate, only applied to the KE. Where will KE users go to find this information now? We shouldn't remove the topic until we have a plan to replace it with the correct content. In short, this PR is only half of the necessary fix. 

I wrote this content based on https://github.com/astronomer/docs/issues/2025 and https://support.astronomer.io/hc/en-us/articles/14968015052179. Tagging @wolfier on this PR to review the content and see if we have evolved our recommendations on this since.

**Comment by jwitz** [comment]
@manmeetkaur These setups are not the exact same thing, because one uses the KPO and one doesn't. We're about to remove instructions that, even if they were inaccurate, only applied to the KE. Where will KE users go to find this information now? We shouldn't remove the topic until we have a plan to replace it with the correct content. In short, this PR is only half of the necessary fix.

I wrote this content based on https://github.com/astronomer/docs/issues/2025 and https://support.astronomer.io/hc/en-us/articles/14968015052179. Tagging @wolfier on this PR to review the content and see if we have evolved our recommendations on this since.

**Comment by manmeetkaur** [comment]
@jwitz I am not sure you understand, let me try and summarize one last time: 

- the section being removed is **does not work**
- **no special instructions are required** for KPO on KE
- `os.environ` **does not work** with KPO for secret env variables set on Astro, what works with KPO is already in the KPO doc.

**Comment by manmeetkaur** [comment]
@jwitz the content in this doc that we are changing and https://github.com/astronomer/docs/issues/2025 is totally different. :)

**Comment by jwitz** [comment]
I understand all of that @manmeetkaur. What I am saying is that we previously identified a need to tell users how to use environment variable secrets with the Kubernetes Executor. That need is no longer being met. Even if the content is just "You can use `os.environ` without further configuration", we should say that here instead of deleting the topic entirely. 

As for your second point, the exact topic you're removing was added in the PR I linked - the `pod_override` code sample is included in the linked ZenDesk article (see https://github.com/astronomer/docs/pull/2054/files#diff-7524d728cea0757a4c90237e05e0910cb40015d397ff24a3905e40f8dbe2fee8R188). Please take a closer look and, if needed, notify the team that our KB articles need to be updated.

**Comment by wolfier** [comment]
> the section being removed does not work

@manmeetkaur Can you be more specific about what is not working? Does the DAG fail? Does the pod requested by the Kubernetes Executor not have the secret specific in the `pod_override`? Please provide a recreation of the behaviour.

If the [documentation](https://docs.astronomer.io/astro/kubernetes-executor#mount-secret-environment-variables-to-worker-pods) and [Knowledge Base article](https://support.astronomer.io/hc/en-us/articles/14968015052179) are wrong, we will need to correct it.

---

The section that is slated for removal in this PR is different than the code that enables mounting a secret for a pod defined by the KubernetesPodOperator.

To mount a secret in for a pod defined by the KubernetesPodOperator, a user can specify the following as shown in our documentation on [Use secret environment variables with the KubernetesPodOperator](https://docs.astronomer.io/astro/kubernetespodoperator?tab=PrivateRegistry#use-secret-environment-variables-with-the-kubernetespodoperator).

```python
    secret_env = Secret(deploy_type="env", deploy_target="MY_SECRET", secret="env-secrets", key="MY_SECRET")
    namespace = conf.get("kubernetes", "NAMESPACE")

    k = KubernetesPodOperator(
        namespace=namespace,
        image="ubuntu:16.04",
        cmds=["bash", "-cx"],
        arguments=["echo $MY_SECRET && sleep 150"],
        name="test-name",
        task_id="test-task",
        get_logs=True,
        secrets=[secret_env],
    )
```

To mount a secret in for a pod requested by the KubernetesExecutor, a user can specify the following as shown in our documentation on [Mount secret environment variables to worker Pods](https://docs.astronomer.io/astro/kubernetes-executor#mount-secret-environment-variables-to-worker-pods).

```python
    secret_env = Secret(deploy_type="env", deploy_target="MY_SECRET", secret="env-secrets", key="MY_SECRET")
    namespace = conf.get("kubernetes", "NAMESPACE")

    p = PythonOperator(
        python_callable=print_env,
        op_kwargs={
            "env_name": secret_env.deploy_target
        },
        task_id='test-py-env',
        executor_config={
            "pod_override": k8s.V1Pod(
                spec=k8s.V1PodSpec(
                    containers=[
                        k8s.V1Container(
                            name="base",
                            env=[
                                k8s.V1EnvVar(
                                    name=secret_env.deploy_target,
                                    value_from=k8s.V1EnvVarSource(
                                        secret_key_ref=k8s.V1SecretKeySelector(name=secret_env.secret,
                                                                               key=secret_env.key)
                                    ),
                                )
                            ],
                        )
                    ]
                )
            ),
        }
    )
```

Please indicate clearly why these two use cases are the same thus only needed one documentation section? The goal is the same but the method to do so is different for different environments. One distinct difference is one uses `executor_config` and `pod_override` and the other uses `KubernetesPodOperator` parameters.

**Comment by manmeetkaur** [comment]
@wolfier Apologies for the delay on this. On K8s executor, adding the pod override makes the task go in a hung state ; it just stays in Queued state. 

Let me run my tests on diff deployments and get the logs ready and then we can chat.

**Comment by manmeetkaur** [comment]
@wolfier Here is the update from my testing with logs:

DAG:
```
from airflow import DAG
from kubernetes.client import models as k8s
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator
from airflow.utils.dates import days_ago
from airflow.configuration import conf
from airflow.operators.python import PythonOperator
from airflow.kubernetes.secret import Secret


def print_env(env_name):
    import os
    print(os.environ[env_name])


def print_env_k8s(secrets):
    print(secrets)


with DAG(
    'env_var_k8s',
    schedule_interval='@daily',
    start_date=days_ago(0),
    tags=['example'],
) as dag:

    secret_env = Secret(deploy_type="env", deploy_target="OWNER", secret="env-secrets", key="OWNER")
    namespace = conf.get("kubernetes", "NAMESPACE")

    p_override = PythonOperator(
        python_callable=print_env,
        op_kwargs={
            "env_name": secret_env.deploy_target
        },
        task_id='test-py-env-override',
        executor_config={
            "pod_override": k8s.V1Pod(
                spec=k8s.V1PodSpec(
                    containers=[
                        k8s.V1Container(
                            name="base",
                            env=[
                                k8s.V1EnvVar(
                                    name=secret_env.deploy_target,
                                    value_from=k8s.V1EnvVarSource(
                                        secret_key_ref=k8s.V1SecretKeySelector(name=secret_env.secret,
                                                                               key=secret_env.key)
                                    ),
                                )
                            ],
                        )
                    ]
                )
            ),
        }
    )

    regular_p = PythonOperator(
        task_id='test-py-env-regular',
        python_callable=print_env,
        op_kwargs={
            "env_name": "OWNER"
        }
    )

    k8s_p = PythonOperator(
        task_id='test-py-env-k8s',
        python_callable=print_env_k8s,
        op_kwargs={
            "secrets": secret_env
        }
    )

    k_env_secrets = KubernetesPodOperator(
        namespace=namespace,
        image="ubuntu:16.04",
        cmds=["bash", "-cx"],
        arguments=["echo $OWNER"],
        name="test-name",
        task_id="test-kpo",
        get_logs=True,
        secrets=[secret_env],
    )

    regular_k = KubernetesPodOperator(
        namespace=namespace,
        image="ubuntu:16.04",
        cmds=["bash", "-cx"],
        arguments=["echo $OWNER"],
        name="test-name",
        task_id="test-kpo-regular",
        get_logs=True,
    )

    p_override >> regular_p >> k8s_p
    k_env_secrets >> regular_k

```

**Problem Statement:** Environment variables, secret or otherwise, are already available to Worker Pods on K8s executor - no override necessary.

**Setup**: 
Create two environment variables
```
ENVIRONMENT: K8s
OWNER(secret): Manmeet
```

**Test Cases as seen in the DAG**:

1. Read a secret environment variable in PythonOperator using Pod override. (task `p_override`): This was stuck/queued in all my earlier tests in one deployment - it succeeded this test - i was using a diff runtime and a new deployment. Which is weird.
2. Read a secret environment variable in a regular pythonic way. using `os.environ`. (task `regular_p`): **_Success_**
3. Read a secret environment variable in a KPO using `k8s.secret` (task: `k_env_secrets`): **Success**
4. Read a secret environment variable in a KPO using regular bash way (task: `regular_k`): **FAILED**

**Airflow URL:** https://cosmicenergy.astronomer.run/ddjjz9lz/dags/env_var_k8s/grid?dag_run_id=manual__2023-09-13T08%3A22%3A27.943020%2B00%3A00&task_id=test-kpo-regular&tab=logs
**Deployment Variables**: https://cloud.astronomer.io/cl81pguhg3568354syfo88o4kjz/deployments/clm7m9l6l474338fdw0gdjjz9lz/variables

**Summary**: 

If you notice, test case #1 uses `pod_override` to achieve the same thing that test case #2 can do using the regular pythonic way. My point is, if we do not have to do anything special to read secret environment variables in a PythonOperator, then why use `pod_override` and complicate for users. Users can seamlessly just use the environment variables using `os.environ` irrespective of the executor. 

Basically there are no "special instructions" for using environment variables in PythonOperator, where as for KPO there are and they are documented in the KPO page. 

I hope this makes sense @wolfier . Let me know if you want to have a call on this and discuss.

**Comment by manmeetkaur** [comment]
As per discussion between @wolfier and @jwitz putting this on hold, because the code itself is correct and works. But we need better understanding of the benefits of using the current way mentioned in our docs and why not use `os.environ`.

**Comment by wolfier** [comment]
> Read a secret environment variable in PythonOperator using Pod override. (task p_override): This was stuck/queued in all my earlier tests in one deployment - it succeeded this test - i was using a diff runtime and a new deployment. Which is weird.

In the above test case, a user could use `os.environ` to achieve the same outcome in a PythonOperator. Because a PythonOperator is flexible, it allows the use of the `os` python package. 

Lets imagine an operator where the user cannot execute arbitrary python code. An operator that expects certain environment variables to exist in the execution environment. Even a python operator whose `python_callable` is predefined and not alterable by the user could fit this scenario.

The only way for the operator to execute successfully is to mount the expected environment variable to the Kubernetes worker pod. If the desired value already exists as a secret environment variable mounted to the deployment via the platform Variables view, the `pod_override` method seems suitable for this case. On the other hand, the `os.environ` method would not work because the operator is not under the control of the user.

**Comment by manmeetkaur** [comment]
@wolfier Agreed on the use-case given. But isn't the current section gives the wrong impression to the user that the only way they can use it by using `pod_override`. What you have specified is a use-case that should be in our docs, but we should clarify that `os.environ` will work and in case you are not able to edit or extend or inherit the pre-defined python method, then this is an option ? I just feel this is a special case as you described, and then we can add details to clarify it. WDYT ?

**Comment by manmeetkaur** [comment]
@jwitz I have updated this PR from the technical perspective. Even though I still feel the use-case of "mount" is also applicable to Celery and not just K8s and it should be not here on the executor page. But you can take that decision where to put it. Also updated the section in environment variables as per the [discussion](https://astronomer.slack.com/archives/C01UA03TPV1/p1695623526387419?thread_ts=1694550962.494019&cid=C01UA03TPV1) in Slack.
